### DISCUSSION TITLE
location of captured data of drone(s) when Python APIs are used.

### DISCUSSION THREAD

**Discussion by AIKUUM** [microsoft/AirSim#3312] [Support Q&A]
System details: https://user-images.githubusercontent.com/72294070/102069277-90db4a00-3e23-11eb-8f71-d14e10181433.PNG
AirSim version (1.3.1)
Unreal version(4.25)
python version (3.9.1)

Consider the following python API: client.getMultirotorState(vehicle_name="Drone1")
Assume that I use City Environment for simulation. Can someone tell me the location (directory) where the text data of '''client.getMultirotorState(vehicle_name="Drone1")''' gets saved?

**Comment by ahmed-elsaharti** [comment]
Hi @AIKUUM, the API calls **return** the needed data when called. ie: a statement like `x = client.getMultirotorState()` would return the multirotor state and store it in variable `x`. AFAIK the calls do not actually save the data in a directory (unless ofcourse you write some code to store the returned value somewhere)

 Please let me know if this answers your question

**Reply by AIKUUM** [reply]
> Hi @AIKUUM, the API calls **return** the needed data when called. ie: a statement like `x = client.getMultirotorState()` would return the multirotor state and store it in variable `x`. AFAIK the calls do not actually save the data in a directory (unless ofcourse you write some code to store the returned value somewhere)
> 
> Please let me know if this answers your question

Hi @ahmed-elsaharti, thank you for your reply. But I got to know what '''x = client.getMultirotorState()'' means few months ago itself. What I feel is that its impossible that return value of "client.getMultirotorState()" doesn't get stored anywhere in the device. There must be some text file or some other file in some directory where the data gets saved, though its being assigned to another variable. 

I'm sorry, but my question has not been answered yet.

**Reply by ahmed-elsaharti** [reply]
@AIKUUM generally speaking, the value gets stored in the temporary memory (RAM) while the program (the simulation) is being executed. It does not get saved into the drive (ie: there is no directory for it to be saved in).
This is similar to you calling any function that returns a value within a program.
For example: in python, if you call math.sqrt(25) it returns 5. That value of '5' is not really stored on your drive at a location, it gets stored in the temporary memory when the program is being executed. Once the execution is complete (when the program is terminated), it's gone.

Edit: saving the information from an API call (or any variable) to a file can be done by writing code to create a file and dump the information needed into it. This is NOT done by default when a program is executed though.
Also, [this](https://www.quora.com/Why-are-variables-of-C-stored-in-RAM-and-not-in-the-hard-disk#:~:text=If%20you're%20storing%20something,small%20%2C%20temporary%20pieces%20of%20data%20.&text=Besides%20%2C%20modern%20operating%20systems%20use%20swap%20files%20and%20virtual%20memory%20.) might help clarify things

**Reply by AIKUUM** [reply]
> @AIKUUM generally speaking, the value gets stored in the temporary memory (RAM) while the program (the simulation) is being executed. It does not get saved into the drive (ie: there is no directory for it to be saved in).
> This is similar to you calling any function that returns a value within a program.
> For example: in python, if you call math.sqrt(25) it returns 5. That value of '5' is not really stored on your drive at a location, it gets stored in the temporary memory when the program is being executed. Once the execution is complete (when the program is terminated), it's gone.
> 
> Edit: saving the information from an API call (or any variable) to a file can be done by writing code to create a file and dump the information needed into it. This is NOT done by default when a program is executed though.
> Also, [this](https://www.quora.com/Why-are-variables-of-C-stored-in-RAM-and-not-in-the-hard-disk#:~:text=If%20you're%20storing%20something,small%20%2C%20temporary%20pieces%20of%20data%20.&text=Besides%20%2C%20modern%20operating%20systems%20use%20swap%20files%20and%20virtual%20memory%20.) might help clarify things

Hi @ahmed-elsaharti, thank you for your reply. But the text which you have mentioned in Edit part, I was hoping that to happen by default, not me writing a separate code for that. This is because, I wanted to regard the drone in the simulation as sender, and another drone in the simulator, or my desktop as receiver. So for my task, if i regard my desktop as receiver, i had already written a piece of code for storing the text data and images captured by the drone to my desktop; but I dont have proof that drone 1(sender) is actually capturing some data.

**Reply by ahmed-elsaharti** [reply]
Sadly, that does not happen by default. You would still have to write some code to fetch the data from the temporary memory and write it into a file for this to happen.
So (if i understood correctly) if your sender is expected to send data through a text file, you will need to have something create that text file and save data from the APIs into it for the receiver to read it

**Reply by AIKUUM** [reply]
> Sadly, that does not happen by default. You would still have to write some code to fetch the data from the temporary memory and write it into a file for this to happen.
> So (if i understood correctly) if your sender is expected to send data through a text file, you will need to have something create that text file and save data from the APIs into it for the receiver to read it

Hi @ahmed-elsaharti, Actually I want communication of multimedia data to happen between sender (drone 1 in simulation) and receiver (drone 2 in simulation/ my desktop). So I cant simply fetch data from RAM. If corresponding communication nodes would get created for drones in simulation, it would have helped.

**Reply by ahmed-elsaharti** [reply]
Is drone 1 being simulated on a different machine?

**Reply by AIKUUM** [reply]
> Is drone 1 being simulated on a different machine?

Hi @ahmed-elsaharti, both are being simulated on same machine. It appears as if you have some better idea to suggest for communication to happen between 2 drones. Can u suggest?

**Reply by ahmed-elsaharti** [reply]
Since both drones are being simulated on the same machine, to have multimedia go from a 'sender' to a 'receiver' the easiest way, as mentioned above, would be to have the sender save the image/text to a file, and have the receiver read the file (since they're on the same machine). Also, since it's the same machine, you **can** access data from the RAM through both the sending and/or receiving program/script.
To be completely honest I'm failing to understand where the issue is here.

**Reply by AIKUUM** [reply]
> Since both drones are being simulated on the same machine, to have multimedia go from a 'sender' to a 'receiver' the easiest way, as mentioned above, would be to have the sender save the image/text to a file, and have the receiver read the file (since they're on the same machine). Also, since it's the same machine, you **can** access data from the RAM through both the sending and/or receiving program/script.
> To be completely honest I'm failing to understand where the issue is here.

Hi @ahmed-elsaharti, issue is that how can I regard whatever you have told as a multimedia communication between a UAV and ground station/drone 2? I cant simply fetch data and simply save it in some directory, how will this justify communication between a UAV and drone 2/ground station? Where are the communication nodes getting created ??

Its something like I cant simply fetch data from simulation, and save it in 10 different directories, and claim that I have 1 sender node and 10 receiver nodes in the simulation.

If drone flights would be stable with PX4, and/or if PX4 would work even with precompiled AirSim binaries, I could have used 2 instances of PX4 in its console, 2 drones would be simulated in unreal engine, I would have transmitted some multimedia data between 2 instances and could have probably regarded it as communication between a sender node and receiver node.

**Reply by ahmed-elsaharti** [reply]
I'm slightly confused here 🤔

Your original question (and first 3-4 replies) conveyed that your main concern was to get the information saved into a:
> location (directory) 

and were
> hoping that to happen by default, not me writing a separate code for that. 

to be used in the context of:
> This is because, I wanted to regard the drone in the simulation as sender, and another drone in the simulator, or my desktop as receiver.

and now you're saying that using this methodology is not valid and that you
> cant simply fetch data and simply save it in some directory
_________________________
Regardless of all that, if an `inter-drone communication` system is to be simulated, you would have to write both 'nodes' anyway. The protocol of communication can be anything and would be dictated by the nodes you write.
_________________________
Lastly, to answer the initial question:
> Can someone tell me the location (directory) where the text data of '''client.getMultirotorState(vehicle_name="Drone1")''' gets saved?

As previously mentioned,
> the API calls return the needed data when called. AFAIK the calls do not actually save the data in a directory.

ie: they do not get saved in a directory, they get saved in the RAM during simulation execution.

Hope this helps 🤷

**Reply by AIKUUM** [reply]
Hi @ahmed-elsaharti, I will clarify everything.

My actual goal is multimedia communication between a sender (drone 1 in the simulation) and receiver (drone 2 in the simulation/my desktop). But a contributor of AirSim had told that communication model is out of scope for AirSim. But I dint want to disturb anyone with this doubt in this discussion.

So I just wanted to know if data captured by a drone (images and text) gets stored in any of the directories. You said that doesn't happen. But your suggestion was to simply save the captured data in some directory which I'm already doing at the receiver end (to regard my desktop as ground station). But you got confused, and thats the reason why I'm mentioning my actual goal.

Morover, can you tell me what you mean by 'writing nodes'? Do you mean configuring the corresponding nodes (sender and receiver) in python script? If I do that, will the communication be simulated like atleast 50% like real time inter-drone communication. What I mean is with increase in the displacement between 2 drones, or between a drone and ground station, will there be increase in packet loss?

Please let me know if you have more doubts.
